python/ray/experimental/queue.py

```diff
@@ -45,70 +43,49 @@ def full(self):
+        if not block:
+            try:
+                ray.get(self.actor.put_nowait.remote(item))
+            except asyncio.QueueFull:
                 raise Full
-        elif timeout is None:
-            # Polling
-            # Use a not_full condition variable or promise?
-            while not ray.get(self.actor.put.remote(item)):
-                # Consider adding time.sleep here
-                pass
-        elif timeout < 0:
-            raise ValueError("'timeout' must be a non-negative number")
         else:
-            endtime = time.time() + timeout
-            # Polling
-            # Use a condition variable or switch to promise?
-            success = False
-            while not success and time.time() < endtime:
-                success = ray.get(self.actor.put.remote(item))
-            if not success:
-                raise Full
+            if timeout is not None and timeout < 0:
+                raise ValueError("'timeout' must be a non-negative number")
+            else:
+                ray.get(self.actor.put.remote(item, timeout))
 
     def get(self, block=True, timeout=None):
         """Gets an item from the queue.
 
-        Uses polling if block=True, so there is no guarantee of order if
-        multiple consumers get from the same empty queue.
+        There is no guarantee of order if multiple consumers get from the
+        same empty queue.
 
         Returns:
             The next item in the queue.
 
         Raises:
             Empty if the queue is empty and blocking is False.
+            Empty if the queue is empty, blocking is True, and it timed out.
+            ValueError if timeout is negative.
         """
         if not block:
-            success, item = ray.get(self.actor.get.remote())
-            if not success:
+            try:
+                return ray.get(self.actor.get_nowait.remote())
+            except asyncio.QueueEmpty:
                 raise Empty
-        elif timeout is None:
-            # Polling
-            # Use a not_empty condition variable or return a promise?
-            success, item = ray.get(self.actor.get.remote())
-            while not success:
-                # Consider adding time.sleep here
-                success, item = ray.get(self.actor.get.remote())
-        elif timeout < 0:
-            raise ValueError("'timeout' must be a non-negative number")
         else:
-            endtime = time.time() + timeout
-            # Polling
-            # Use a not_full condition variable or return a promise?
-            success = False
-            while not success and time.time() < endtime:
-                success, item = ray.get(self.actor.get.remote())
-            if not success:
-                raise Empty
-        return item
+            if timeout is not None and timeout < 0:
+                raise ValueError("'timeout' must be a non-negative number")
+            else:
+                return ray.get(self.actor.get.remote(timeout))
```

Comment:
Same thing here! I think the timeout could cause an item to be popped from the queue and ignored without calling `ray.cancel`